### DISCUSSION TITLE
Versioning API support

### DISCUSSION THREAD

**Discussion by lraulet** [versity/versitygw#602] [Q&A]
Hi,
I'm currently testing versitygw and it works well for now ; still evaluating the solution.
I'm using the Posix backend.
I saw that Versioning API (bucket and objects) are not supported.
Is there any plan to support one or both in the futur ?
If the answer is yes and you already have ideas on how to do this, I (and my team) may be able to help adding this feature.

**Comment by benmcclelland** [comment]
@lraulet Thank you for the offer. We don't have a specific design for how to support this yet. But we have been interested in exploring ideas for how we might enable this.

The main issue is that posix filesystems don't inherently support file versioning in any generic way. So I'm guessing that we would end up storing copies of files in the filesystem as versions. The struggle is where to put these files. We need to be careful not to have any name conflicts with objects that might be uploaded. For example:
PUT bucket=mybucket object=dir/myobject
would end up in the filesystem as <root>/mybucket/dir/myobject. So we might be tempted to append versions to the name, such as:
<root>/mybucket/dir/myobject.1, but then what happens if someone needs to upload an object called "dir/myobject.1".
We could maybe require all objects always have a version suffix, so the above would result in the object "dir/myobject.1.0", but then this breaks the case where people may want to copy files into the filesystem and have them show up as objects through the gateway.
So I'm not very satisfied with any object name manipulation.

Another alternative might be to have a directory at the bucket level that sort of shadows the bucket that would store the non-primary versions. This could work, but we would need to make sure these wouldnt conflict with buckets that someone might try to create.

And the final alternative that I have considered would be to make use of specific filesystem features for filesystems that might support natively support versioning. See [Versioning_file_system](https://en.wikipedia.org/wiki/Versioning_file_system#Linux). This would be the cleanest approach, but is probably a non-starter seeing as how none of the mainstream filesystems support this.

Since none of these options were feeling very great to me, I was sort of hoping that someone else might have some better ideas for a better design.

**Comment by benmcclelland** [comment]
Opening issue #678 to discuss design/implementation ideas for adding versioning support.
